File: backend/Models/Book.cs
namespace Backend.Models;

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

public class Book : BaseModel
{
    [MinLength(1)]
    [MaxLength(250)]
    public string Title { get; set; } = null!;

    [MaxLength(500)]
    public string? Description { get; set; }

    public ICollection<Category> Categories { get; set; } = null!;
    public ICollection<Author> Authors { get; set; } = null!;
    public ICollection<Copy> Copies { get; set; } = null!;

    [NotMapped]
    [JsonIgnore]
    public ICollection<Publisher>? Publishers 
    { 
        get => Copies?.Select(c => c.Publisher).Distinct().ToList(); 
    }

    [NotMapped]
    public int CopiesAvailable 
    { 
        get => Copies == null ? 0 : Copies.Where(c => c.IsAvailable).Count();
    }

}

File: frontend/src/types/book.ts
import { Author } from "./author";
import { Base } from "./base";
import { Category } from "./category";
import { Copy } from "./copy";
import { Publisher } from "./publisher";

export interface Book extends Base {
    title: string,
    description: string | null,
    authors: Author[] | null,
    publishers: Publisher[] | null
    categories: Category[] | null,
    copies: Copy[] | null,
    copiesAvailable: number | null,
};

export type CreateBook = Omit<Book, "id" | "categories" | "authors" | "publishers" | "copies" | "copiesAvailable">;

export type AddToBook = {
    id: number,
    addId: number
} 